app/scripts/modules/core/src/artifact/ArtifactReferenceService.ts

```diff
@@ -0,0 +1,32 @@
+import { module } from 'angular';
+
+interface Walker {
+  (instance: any, visitor: any): void
+}
+
+interface Reference {
+  category: string;
+  walker: Walker;
+}
+
+export class ArtifactReferenceServiceProvider {
+  private references: Reference[] = [];
+
+  public $get() {
+    return this;
+  }
+
+  public registerReference(category: string, walker: any) {
+    this.references.push({ category, walker });
+  }
+
+  public walkReferences(category: string, instance: any, visitor: any) {
```

Comment:
I don't see a super strong reason for each `visitor` to be dynamic, rather than statically defined on the `ArtifactReferenceService` itself (especially since you only have one `visitor` right now), but it's up to you. You could have each `Walker` return a `[object, key]` tuple, and a method on `ArtifactReferenceService` could pull the key from the object.

Reply:
`IWalker`s now return a list of paths into the stage and the dynamic visitor has been removed.  The register API feels less complicated in the trivial case.  The google clone stage still has to do a bunch of work to construct its list of paths though.